File: server/stepmodel.js
// stepmodel.js
// Mini-red neuronal (MLP) para STEP-SELECTION: dado el contexto musical de una
// nota (instrumento, tono, posicion en el beat, rapidez) MAS la historia
// reciente de carriles y el estado del pie, predice una distribucion sobre los
// carriles. Aprende FORMAS DE PATRON (escaleras, crossovers, candles, boxes,
// alternancia) que un scorer nota-a-nota no produce solo.
//
// - Inferencia en JS PURO (sin dependencias, corre en el motor Node del juego).
// - Los pesos se empaquetan con el juego (server/model/step-model-*.json), asi
//   el usuario NO descarga nada.
// - El modelo se MEZCLA con el mapeo musical heuristico (no lo reemplaza): el
//   heuristico mantiene tono/instrumento/jugabilidad; el modelo aporta la forma
//   del patron de varios pasos. Si no hay pesos, todo sigue funcionando igual.

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const MODEL_DIR = path.join(__dirname, "model");

// ---------- Features (debe coincidir con el entrenador y con StepModel.kt) ----------
// Vector de entrada para predecir el carril de UNA nota:
//   pitch (1)
//   voz one-hot: kick, hat, cymbal, melody (4)
//   onDownbeat (1), beatInBar (1), strong (1), reach (1), looseness (1), prevJump (1)
//   foot: isLeft, isRight (2)
//   historia de carriles one-hot: last1..last5  (5*L)
// Total = 13 + 5*L.  (L=5 -> 38 ; L=4 -> 33)
export function buildFeatures(ctx, laneCount) {
  const L = laneCount;
  const f = new Float32Array(13 + 5 * L);
  let i = 0;
  f[i++] = clamp01(ctx.pitch);
  const v = ctx.voice;
  f[i++] = v === "kick" ? 1 : 0;
  f[i++] = v === "hat" ? 1 : 0;
  f[i++] = v === "cymbal" ? 1 : 0;
  f[i++] = (v === "melody" || !v) ? 1 : 0;
  f[i++] = ctx.onDownbeat ? 1 : 0;
  f[i++] = clamp01(ctx.beatInBar);      // posicion dentro del compas (0..1)
  f[i++] = clamp01(ctx.strong);
  f[i++] = clamp01(ctx.reach);          // 1 = nota muy rapida, 0 = lenta
  f[i++] = clamp01(ctx.looseness);
  f[i++] = ctx.prevJump ? 1 : 0;        // la nota previa fue un jump/acorde
  f[i++] = ctx.foot < 0 ? 1 : 0;        // ultimo pie izquierdo
  f[i++] = ctx.foot > 0 ? 1 : 0;        // ultimo pie derecho
  // Historia de los ultimos 5 carriles (trigrama ampliado a 5 para capturar
  // formas largas: escaleras de 5, boxes, crossovers extendidos).
  const hist = [ctx.lastLane, ctx.last2Lane, ctx.last3Lane, ctx.last4Lane, ctx.last5Lane];
  for (let hpos = 0; hpos < 5; hpos++) {
    const lane = hist[hpos];
    for (let k = 0; k < L; k++) f[i++] = lane === k ? 1 : 0;
  }
  return f;
}

function clamp01(x) { x = +x; return x < 0 ? 0 : x > 1 ? 1 : (isNaN(x) ? 0 : x); }

// ---------- MLP: forward pass (2 capas ocultas ReLU, salida softmax) ----------
// weights = { inDim, hidden, hidden2, out, W1,b1, W2,b2, W3,b3 }.
// Compatible hacia atras: si no hay hidden2/W3, usa 1 sola capa (W2 -> salida).
function forward(weights, x) {
  const { inDim, hidden, hidden2, out, W1, b1, W2, b2, W3, b3 } = weights;
  const h1 = new Float32Array(hidden);
  for (let j = 0; j < hidden; j++) {
    let s = b1[j]; const base = j * inDim;
    for (let k = 0; k < inDim; k++) s += W1[base + k] * x[k];
    h1[j] = s > 0 ? s : 0;               // ReLU
  }
  const logits = new Float32Array(out);
  if (hidden2 && W3) {
    const h2 = new Float32Array(hidden2);
    for (let o2 = 0; o2 < hidden2; o2++) {
      let s = b2[o2]; const base = o2 * hidden;
      for (let j = 0; j < hidden; j++) s += W2[base + j] * h1[j];
      h2[o2] = s > 0 ? s : 0;            // ReLU
    }
    for (let o = 0; o < out; o++) { let s = b3[o]; const base = o * hidden2; for (let o2 = 0; o2 < hidden2; o2++) s += W3[base + o2] * h2[o2]; logits[o] = s; }
  } else {
    for (let o = 0; o < out; o++) { let s = b2[o]; const base = o * hidden; for (let j = 0; j < hidden; j++) s += W2[base + j] * h1[j]; logits[o] = s; }
  }
  // softmax (devolvemos log-probabilidades para mezclar como prior aditivo)
  let mx = -Infinity; for (let o = 0; o < out; o++) if (logits[o] > mx) mx = logits[o];
  let sum = 0;
  for (let o = 0; o < out; o++) { logits[o] = Math.exp(logits[o] - mx); sum += logits[o]; }
  const logp = new Float32Array(out);
  for (let o = 0; o < out; o++) logp[o] = Math.log((logits[o] / sum) + 1e-9);
  return logp;
}

// ---------- Carga perezosa de pesos por laneCount ----------
const _cache = new Map();
function loadWeights(laneCount) {
  if (_cache.has(laneCount)) return _cache.get(laneCount);
  let w = null;
  try {
    const file = path.join(MODEL_DIR, `step-model-${laneCount}.json`);
    if (fs.existsSync(file)) {
      const raw = JSON.parse(fs.readFileSync(file, "utf8"));
      w = {
        inDim: raw.inDim, hidden: raw.hidden, hidden2: raw.hidden2 || 0, out: raw.out,
        W1: Float32Array.from(raw.W1), b1: Float32Array.from(raw.b1),
        W2: Float32Array.from(raw.W2), b2: Float32Array.from(raw.b2),
        W3: raw.W3 ? Float32Array.from(raw.W3) : null, b3: raw.b3 ? Float32Array.from(raw.b3) : null,
      };
    }
  } catch (_) { w = null; }
  _cache.set(laneCount, w);
  return w;
}

export function hasStepModel(laneCount) { return !!loadWeights(laneCount); }

// Invalida la cache de pesos: tras reentrenar, la proxima prediccion releera
// los pesos nuevos del disco. Asi el modelo recien entrenado entra en uso sin
// reiniciar el motor.
export function reloadStepModel() { _cache.clear(); }

// Devuelve un array de log-probabilidades por carril (length=laneCount) o null
// si no hay modelo disponible para ese laneCount.
export function predictLaneLogits(ctx, laneCount) {
  const w = loadWeights(laneCount);
  if (!w) return null;
  const x = buildFeatures(ctx, laneCount);
  if (x.length !== w.inDim) return null;     // incompatibilidad de features
  return forward(w, x);
}

File: android/app/src/main/java/com/rhythmdance/app/game/StepModel.kt
package com.rhythmdance.app.game

import android.content.Context
import org.json.JSONObject
import kotlin.math.exp
import kotlin.math.ln

// Mini-red neuronal (MLP) para STEP-SELECTION, portada 1:1 desde la versión JS
// (server/stepmodel.js). Dado el contexto musical de una nota + la historia
// reciente de carriles, predice log-probabilidades por carril. Aporta la FORMA
// del patrón (escaleras, crossovers, boxes, alternancia) que un scorer
// nota-a-nota no produce solo. Se MEZCLA con la heurística (no la reemplaza).
//
// Pesos empaquetados en assets/model/step-model-{5,4}.json (no se descarga nada).
// Arquitectura: entrada -> capa oculta (ReLU) -> salida (softmax -> log-probs).
class StepModel private constructor(
    private val inDim: Int,
    private val hidden: Int,
    private val hidden2: Int,
    private val out: Int,
    private val w1: FloatArray, private val b1: FloatArray,
    private val w2: FloatArray, private val b2: FloatArray,
    private val w3: FloatArray?, private val b3: FloatArray?,
) {
    // Contexto de UNA nota para construir el vector de features.
    class Ctx(
        val pitch: Double, val voice: String?, val onDownbeat: Boolean,
        val beatInBar: Double, val strong: Double, val reach: Double,
        val looseness: Double, val prevJump: Boolean, val foot: Int,
        val lastLane: Int, val last2Lane: Int, val last3Lane: Int,
        val last4Lane: Int = -1, val last5Lane: Int = -1,
    )

    companion object {
        private val cache = HashMap<Int, StepModel?>()

        // Carga perezosa de pesos por laneCount. Prefiere el modelo REENTRENADO
        // en el teléfono (filesDir/model) sobre el empaquetado (assets/model).
        fun load(context: Context, laneCount: Int): StepModel? {
            if (cache.containsKey(laneCount)) return cache[laneCount]
            var model: StepModel? = null
            val txt = readModelJson(context, laneCount)
            if (txt != null) {
                try {
                    val j = JSONObject(txt)
                    model = StepModel(
                        j.getInt("inDim"), j.getInt("hidden"), j.optInt("hidden2", 0), j.getInt("out"),
                        toFloatArray(j.getJSONArray("W1")), toFloatArray(j.getJSONArray("b1")),
                        toFloatArray(j.getJSONArray("W2")), toFloatArray(j.getJSONArray("b2")),
                        if (j.has("W3")) toFloatArray(j.getJSONArray("W3")) else null,
                        if (j.has("b3")) toFloatArray(j.getJSONArray("b3")) else null,
                    )
                } catch (e: Exception) { model = null }
            }
            cache[laneCount] = model
            return model
        }

        // Texto del modelo: filesDir/model (reentrenado) o assets/model (base).
        fun readModelJson(context: Context, laneCount: Int): String? {
            val f = java.io.File(java.io.File(context.filesDir, "model"), "step-model-$laneCount.json")
            if (f.exists()) { try { return f.readText() } catch (_: Exception) {} }
            return try { context.assets.open("model/step-model-$laneCount.json").bufferedReader().use { it.readText() } }
            catch (e: Exception) { null }
        }

        // Invalida la caché para releer pesos tras reentrenar en el teléfono.
        fun clearCache() { cache.clear() }

        private fun toFloatArray(arr: org.json.JSONArray): FloatArray {
            val out = FloatArray(arr.length())
            for (i in 0 until arr.length()) out[i] = arr.getDouble(i).toFloat()
            return out
        }

        private fun clamp01(x: Double): Float { if (x.isNaN()) return 0f; return if (x < 0) 0f else if (x > 1) 1f else x.toFloat() }

        // Vector de features (compartido por inferencia y entrenamiento en device).
        // Debe coincidir EXACTO con buildFeatures de stepmodel.js.
        fun features(c: Ctx, L: Int): FloatArray {
            val f = FloatArray(13 + 5 * L)
            var i = 0
            f[i++] = clamp01(c.pitch)
            f[i++] = if (c.voice == "kick") 1f else 0f
            f[i++] = if (c.voice == "hat") 1f else 0f
            f[i++] = if (c.voice == "cymbal") 1f else 0f
            f[i++] = if (c.voice == "melody" || c.voice == null) 1f else 0f
            f[i++] = if (c.onDownbeat) 1f else 0f
            f[i++] = clamp01(c.beatInBar)
            f[i++] = clamp01(c.strong)
            f[i++] = clamp01(c.reach)
            f[i++] = clamp01(c.looseness)
            f[i++] = if (c.prevJump) 1f else 0f
            f[i++] = if (c.foot < 0) 1f else 0f
            f[i++] = if (c.foot > 0) 1f else 0f
            val hist = intArrayOf(c.lastLane, c.last2Lane, c.last3Lane, c.last4Lane, c.last5Lane)
            for (hpos in 0 until 5) { val lane = hist[hpos]; for (k in 0 until L) f[i++] = if (lane == k) 1f else 0f }
            return f
        }
    }

    // Forward pass -> log-probabilidades por carril (length = out).
    // 2 capas ocultas ReLU si hay W3; si no, 1 capa (compatibilidad).
    fun predictLogits(c: Ctx, laneCount: Int): FloatArray? {
        val x = features(c, laneCount)
        if (x.size != inDim) return null
        val h1 = FloatArray(hidden)
        for (jn in 0 until hidden) {
            var s = b1[jn]; val base = jn * inDim
            for (k in 0 until inDim) s += w1[base + k] * x[k]
            h1[jn] = if (s > 0f) s else 0f
        }
        val logits = FloatArray(out)
        if (hidden2 > 0 && w3 != null && b3 != null) {
            val h2 = FloatArray(hidden2)
            for (o2 in 0 until hidden2) {
                var s = b2[o2]; val base = o2 * hidden
                for (jn in 0 until hidden) s += w2[base + jn] * h1[jn]
                h2[o2] = if (s > 0f) s else 0f
            }
            for (o in 0 until out) { var s = b3[o]; val base = o * hidden2; for (o2 in 0 until hidden2) s += w3[base + o2] * h2[o2]; logits[o] = s }
        } else {
            for (o in 0 until out) { var s = b2[o]; val base = o * hidden; for (jn in 0 until hidden) s += w2[base + jn] * h1[jn]; logits[o] = s }
        }
        var mx = Float.NEGATIVE_INFINITY
        for (o in 0 until out) if (logits[o] > mx) mx = logits[o]
        var sum = 0f
        for (o in 0 until out) { logits[o] = exp(logits[o] - mx); sum += logits[o] }
        val logp = FloatArray(out)
        for (o in 0 until out) logp[o] = ln((logits[o] / sum) + 1e-9f)
        return logp
    }
}
